sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/constraints/Constraint.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog.constraints;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+import org.apache.spark.sql.connector.expressions.filter.Predicate;
+
+/**
+ * A constraint that defines valid states of data in a table.
+ *
+ * @since 4.1.0
+ */
+@Evolving
+public interface Constraint {
+  /**
+   * Returns the name of this constraint.
+   */
+  String name();
+
+  /**
+   * Returns the definition of this constraint in DDL format.
+   */
+  String toDDL();
+
+  /**
+   * Returns the state of this constraint.
+   */
+  ConstraintState state();
+
+  /**
+   * Creates a CHECK constraint with a search condition defined in SQL (Spark SQL dialect).
+   *
+   * @param name the constraint name
+   * @param sql the SQL representation of the search condition (Spark SQL dialect)
+   * @param state the constraint state
+   * @return a CHECK constraint with the provided configuration
+   */
+  static Check check(String name, String sql, ConstraintState state) {
+    return new Check(name, sql, null /* no predicate */, state);
+  }
+
+  /**
+   * Creates a CHECK constraint with a search condition defined by a {@link Predicate predicate}.
+   *
+   * @param name the constraint name
+   * @param predicate the search condition
+   * @param state the constraint state
+   * @return a CHECK constraint with the provided configuration
+   */
+  static Check check(String name, Predicate predicate, ConstraintState state) {
+    return new Check(name, null /* no SQL */, predicate, state);
+  }
+
+  /**
+   * Creates a CHECK constraint with a search condition defined in SQL (Spark SQL dialect) and
+   * by {@link Predicate} (if the SQL representation can be converted into a supported expression).
+   * The SQL string and predicate must be equivalent.
+   *
+   * @param name the constraint name
+   * @param sql the SQL representation of the search condition (Spark SQL dialect)
```

Comment:
>  sql the SQL representation of the search condition (Spark SQL dialect)

[doubt] How are we planning to handle cases in lets say an inbuilt function is introduced at certain spark version and is referenced in check but the enforcer is at a lower version of spark sql ? 

For ex: 
> ALTER TABLE tbl ADD CONSTRAINT my_condition CHECK (func_spark_4(x) < 20);

Now the writer / enforcer is in  spark_3 and doesn't know func_spark_4

Reply:
That would mean the older version of Spark would not be able to parse/bind the expression and the write will fail. I think that's an acceptable behavior and is better than leading to silent constraint violation.